contracts/BoostedMasterChefJoe.sol

```diff
@@ -0,0 +1,480 @@
+    ) external onlyOwner {
+        PoolInfo memory pool = poolInfo[_pid];
+        totalAllocPoint = totalAllocPoint.sub(poolInfo[_pid].allocPoint).add(_allocPoint);
+        pool.allocPoint = _allocPoint;
+        if (_overwrite) {
+            if (address(_rewarder) != address(0)) {
+                _rewarder.onJoeReward(address(0), 0);
+            }
+            pool.rewarder = _rewarder;
+        }
+        poolInfo[_pid] = pool;
+        emit Set(_pid, _allocPoint, _overwrite ? _rewarder : pool.rewarder, _overwrite);
+    }
+
+    /// @notice View function to see pending JOE on frontend
+    /// @param _pid The index of the pool. See `poolInfo`
+    /// @param _user Address of user
+    /// @return pendingJoe JOE reward for a given user.
+    /// @return bonusTokenAddress The address of the bonus reward.
+    /// @return bonusTokenSymbol The symbol of the bonus token.
+    /// @return pendingBonusToken The amount of bonus rewards pending.
+    function pendingTokens(uint256 _pid, address _user)
+        external
+        view
+        returns (
+            uint256 pendingJoe,
+            address bonusTokenAddress,
+            string memory bonusTokenSymbol,
+            uint256 pendingBonusToken
+        )
+    {
+        PoolInfo memory pool = poolInfo[_pid];
+        UserInfo storage user = userInfo[_pid][_user];
+        uint256 accJoePerShare = pool.accJoePerShare;
+
+        if (block.timestamp > pool.lastRewardTimestamp && pool.totalBoostedAmount != 0) {
+            uint256 secondsElapsed = block.timestamp.sub(pool.lastRewardTimestamp);
+            uint256 joeReward = secondsElapsed.mul(joePerSec()).mul(pool.allocPoint).div(totalAllocPoint);
+            accJoePerShare = accJoePerShare.add(joeReward.mul(ACC_TOKEN_PRECISION).div(pool.totalBoostedAmount));
+        }
+
+        uint256 userLiquidity = getBoostedLiquidity(_pid, _user);
+        pendingJoe = userLiquidity.mul(accJoePerShare).div(ACC_TOKEN_PRECISION).sub(user.rewardDebt).add(
+            claimableJoe[_pid][_user]
+        );
+
+        // If it's a double reward farm, we return info about the bonus token
+        if (address(pool.rewarder) != address(0)) {
+            bonusTokenAddress = address(pool.rewarder.rewardToken());
+            bonusTokenSymbol = IERC20(pool.rewarder.rewardToken()).safeSymbol();
+            pendingBonusToken = pool.rewarder.pendingTokens(_user);
+        }
+    }
+
+    /// @notice Update reward variables for all pools. Be careful of gas spending!
+    /// @param _pids Pool IDs of all to be updated. Make sure to update all active pools
+    function massUpdatePools(uint256[] calldata _pids) external {
+        uint256 len = _pids.length;
+        for (uint256 i = 0; i < len; ++i) {
+            updatePool(_pids[i]);
+        }
+    }
+
+    /// @notice Calculates and returns the `amount` of JOE per second
+    /// @return _amount The amount of JOE emitted per second
+    function joePerSec() public view returns (uint256 _amount) {
```

Comment:
question: whats our convention on the naming of return variables? asking b/c this return has an underscore but the ones in `pendingTokens` function doesn't so we should probably be consistent

Reply:
Hmm, this is the only function that prefixes the return name with `_`. For the time being I have removed this to make consistent with the rest of the functions. The solidity style guide doesn't mention anything about this but there is an unofficial convention of prefixing local variables with `_` or using it when there would be a name clash with a state variable.

I'm totally up for writing our own style guide in notion for things not covered by the solidity style guide.